onnxruntime/core/providers/cpu/math/quantize_linear_matmul.cc

```diff
@@ -97,14 +96,17 @@ Status QLinearMatMul::Compute(OpKernelContext* ctx) const {
   gemm_params.lda = gemm_shape.K;
   gemm_params.ZeroPointA = *a_offset->template Data<uint8_t>();
   gemm_params.ldb = gemm_shape.N;
-  gemm_params.ZeroPointB = static_cast<const uint8_t*>(b_offset->DataRaw());
   gemm_params.C = gemm_output;
   gemm_params.ldc = gemm_shape.N;
   gemm_params.BIsPacked = bool(packed_b_);
+  gemm_params.PerColumnZeroPoints = !IsScalarOr1ElementVector(b_offset);
 
+  auto b_zp_data = static_cast<const uint8_t*>(b_offset->DataRaw());
   for (size_t i = 0; i < helper.OutputOffsets().size(); i++) {
+    auto b_param_offset = helper.RightOffsets()[i] / helper.K();
```

Comment:
it's strange to have some of the buffer offset given by helper function but others do not.

Reply:
For a Matrix B with shape [A1, A2, A3, K, N], its scale and zero point will be have shape like [A1, A2, A3, 1, N] for per-column quantization. For per-tensor quantization, the scale and zero point is a scalar. That's why per-column has offset but per-tensor doesn't.

Reply:
I thought helper is for batch gemm, and find out where the input for each gemm? Still in this case, you are calculating the position of b zero point buffer outside of helper. So you have to make certain assumption of the layout. I think the benefit of having a helper is to encapsulate all these assumptions in one place ?

Reply:
Just got your point. This is not an assumption. It is validated by IsBQuantParamSupported call above. Yes, we can add all the scale and zero point offset to the helper, but it will make it complicated because the offset of scale/zero can be calculated easily with A/B offset. Won't bother to do that.

Reply:
consider adding comments. and the variable name is kinda confusing to me.

Reply:
Just added the offset support for zero point and scale in MatMulHelper.